### PR TITLE
Added script to invite users from one channel to another

### PR DESCRIPTION
**Pull request by bhurstGH**
Here it is, as-is!

### CODE REVIEW

File: package-lock.json [hunk 1]
```diff
@@ -0,0 +1,48 @@
+{
```
Comment (phoenixcoder) [inline package-lock.json:1]: This file looks auto-generated. Is this file necessary? If it can be generated at runtime by the build command, let's get rid of this.

File: package.json [hunk 2]
```diff
@@ -0,0 +1,18 @@
+{
```
Comment (phoenixcoder) [inline package.json:1]: Same here. Is this necessary?

File: index.js [hunk 3]
```diff
@@ -0,0 +1,100 @@
+#!/usr/bin/env node
+require("dotenv").config();
+const axios = require("axios");
+
+// Store Slack access token in a .env file in the root folder
```
Comment (phoenixcoder) [inline index.js:12]: I recommend writing out the instructions for how this intended to be used in comments. It's better to add it here, than the README since the overall project is to botify it whereas this is a script intended to be run on its own right now.

File: index.js [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+const accessToken = process.env.SLACK_ACCESS_TOKEN;
+
+async function getConversationId(conversation) {
```
Comment (nspilman) [inline index.js:17]: Can we rename Conversation to Channel? Conversation == Slack Channel, no?

File: index.js [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+  // Find the channel with matching conversation name and pull out its ID
+  const conversationId = channels.find(chan => chan.name === conversation).id
+
+  return conversationId
+}
+
+async function getConversationMembers(conversationId) {
+
+  async function queryForMembers(membersCollection = [], nextCursor = '') {
```
Comment (nspilman) [inline index.js:35]: what is nextCursor?

File: index.js [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+      );
+
+      // Looks ugly, but pulling out members and next_cursor
+      let {
+        data: {
+          members,
+          response_metadata: { next_cursor }
+        }
+      } = response;
+
```
Comment (nspilman) [inline index.js:55]: I'm honestly not familiar with the above syntax.

File: index.js (hunk 5 above)
Comment (nspilman) [reply index.js:35]: I also don't know I like having queryForMembers scoped within the getConversationMembers function

**Review by nspilman — COMMENTED** [review]
Looks good to me. I have a few questions about the setup functions, but the inviteMembers function is nice and readable

File: index.js (hunk 4 above)
Comment (bhurstGH) [reply index.js:17]: Mark just pointed out to me that there were comments here. Sorry, I never got notified via email. D'oh!

We absolutely could. The reason it is "Conversation" is because Slack is pushing towards using their Conversation API, which could be a channel or a private DM, etc. I named it line up with the API. But in this context, yes, it could just be channel.

File: index.js (hunk 5 above)
Comment (bhurstGH) [reply index.js:35]: nextCursor is following the language of the response from Slack. The results are paginated, and "next_cursor" is the indication that there is another page of results to poll.

Scoping: Absolutely. This is a hold over from refactoring. There's no need for it now. Though a function in a function and taking advantage of closures in this way isn't uncommon, there's no longer a purpose for it now since I changed how I handled it over iterations. Good catch.

File: index.js (hunk 6 above)
Comment (bhurstGH) [reply index.js:55]: Assuming you are referring to the "Looks ugly, but..." bit, that's nested destructuring, pulling the results from the response object that axios returned. It could be rewritten as:

`let members = response.data.members`
`let next_cursor = response.data.response_metadata.next_cursor`

OR, with normal destructuring instead of nested:

`let { members } = response.data`
`let { next_cursor } = response.data.response_metadata`

Destructuring allows you to pull out a property from an object into a variable of the same name.

File: package.json (hunk 2 above)
Comment (bhurstGH) [reply package.json:1]: package.json definitely does, as it holds the dependency list and npm project data. package-lock.json I am unsure of, but assume NPM creates it for a reason.

Comment (phoenixcoder) [reply package.json:1]: I was looking for a gist to be added here of the output you got on the terminal when you proved this. It's not a big deal, but it's usually courteous to the reviewer and others vetting your changes in the future to have that.

File: index.js [hunk 7]
```diff
@@ -2,6 +2,13 @@
 require("dotenv").config();
 const axios = require("axios");
 
+// Usage from the command line:
+// npm install
+// node index sourceChannel targetChannel
+// e.g. node index general dev-chatter
+
+// Note: Do not include the hashtag in the channel name. Though I did account for stripping the hashtag away, process.argv returns undefined if the hashtag is present and I'm unsure of why.
```
Comment (phoenixcoder) [inline index.js:10]: You also need to indicate how to get a hold of the Slack access token e.g. ping a collaborator to be added as a project collaborator, and then the page to go to after than (or a Slack link telling you how) to copy the access token generated.

**Review by phoenixcoder — CHANGES_REQUESTED** [review]

File: package.json (hunk 2 above)
Comment (bhurstGH) [reply package.json:1]: Certainly!
```
Bretts-MBP:struggle-bus-bot bretthurst$ npm install
npm notice created a lockfile as package-lock.json. You should commit this file.
npm WARN slack-inviteall@1.0.0 No repository field.

added 6 packages from 8 contributors and audited 6 packages in 0.633s
found 0 vulnerabilities
```

It created the package-lock on install, but did give warning and suggest that it gets included in the commit.

File: index.js (hunk 7 above)
Comment (bhurstGH) [reply index.js:10]: Is it appropriate to squeeze that sort of instruction into code comments rather than in onboarding documentation somewhere?

Comment (phoenixcoder) [reply index.js:10]: @bhurstGH if I were reading this for the first time, I'd want that mentioned, if I'm trying to run this. Functionally, that's needed to make this run. So, that part of the process is required. If the link is to just an onboarding document, then that's fine too. However, considering we don't have that document yet (?) we could just tell them the two points above.

File: index.js [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+
+async function getConversationId(conversation) {
+
+  // Trim the pound sign if it was provided in the conversation argument
+  if (conversation.startsWith('#')) {
+    conversation = conversation.slice(1)
+  }
+
+  // Destructuring to pull out the channels result from the response
+  const { data: { channels } } = await axios.get(`https://slack.com/api/conversations.list?token=${accessToken}&pretty=1`)
```
Comment (phoenixcoder) [inline index.js:25]: nit-pick: I really don't like strings like this in the code. It should be a constant with a name documenting what the URL is for.

File: index.js [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+
+  // Find the channel with matching conversation name and pull out its ID
+  const conversationId = channels.find(chan => chan.name === conversation).id
```
Comment (phoenixcoder) [inline index.js:28]: This is pulling out a channel id at the end from the conversations API. So, this method should in fact be named getChannelId().

File: index.js [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+
+  return conversationId
+}
+
+async function getConversationMembers(conversationId) {
```
Comment (phoenixcoder) [inline index.js:33]: See above comment. This should be getChannelMembers() now.

File: index.js [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+
+  async function queryForMembers(membersCollection = [], nextCursor = '') {
+    let allConversationMembers = [];
```
Comment (phoenixcoder) [inline index.js:36]: Same conversation => channel

File: index.js [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+
+    // Base case: members exist, no next cursor, we've retrieved all the members
+    if (membersCollection.length > 0 && !nextCursor) {
```
Comment (phoenixcoder) [inline index.js:39]: membersCollection.length > 0 is extraneous. You should just return whatever list you have. The n-1th recursion should be able to handle an empty list.

Also, your base case will never trigger, if the totalNumberOfMembers % pageSize === 0. This would be a bug.

File: index.js [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+      return membersCollection;
+    }
+
+    try {
+      const response = await axios.get(
+        `https://slack.com/api/conversations.members?token=${accessToken}&channel=${conversationId}${nextCursor}`
```
Comment (phoenixcoder) [inline index.js:45]: Same re. undocumented URL strings.

File: index.js [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+      } = response;
+
+      // Cursor seems to always end with =
+      // Replace with URL encoding, %3D
+      if (next_cursor) {
+        next_cursor = `&cursor=${next_cursor.replace('=', '%3D')}`
+      }
+
+      // Combine member grouping with previous
+      allConversationMembers = [...members, ...membersCollection]
```
Comment (phoenixcoder) [inline index.js:63]: Conversation => channel

File: index.js (hunk 6 above)
Comment (phoenixcoder) [reply index.js:55]: @bhurstGH you should write this in a manner that is most clear to your audience. If Nate can't read it, you may want to consider changing it to the clearer version instead, or argue for it as the standard way to do things.

Given that you say in a comment: "Looks ugly..." It's probably not the clearest way to do things.

File: index.js [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+        data: {
+          members,
+          response_metadata: { next_cursor }
+        }
+      } = response;
+
+      // Cursor seems to always end with =
+      // Replace with URL encoding, %3D
+      if (next_cursor) {
+        next_cursor = `&cursor=${next_cursor.replace('=', '%3D')}`
```
Comment (phoenixcoder) [inline index.js:59]: Are you sure this is the right thing to do? You should never modify cursor tokens...I'm surprised this didn't just break.

File: index.js [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+    return console.log("Please supply a source channel and a target channel")
+  }
+
+  // Retrieve the ID of the source and target channels for member access
+  const sourceId = await getConversationId(sourceChannel)
+  const targetId = await getConversationId(targetChannel)
+
+  // Retrieve member lists for comparison
+  const sourceMembers = await getConversationMembers(sourceId)
+  const targetMembers = await getConversationMembers(targetId)
```
Comment (phoenixcoder) [inline index.js:90]: I think, this await and the ones above it are not taking advantage of concurrency. Basically, each one is waiting on the last promise to finish before running its own code. You might as well get rid of the async function and call it synchronously.

IRC, you need to encapsulate these function call in yet another promise to run concurrently.

Reference: https://tutorialzine.com/2017/07/javascript-async-await-explained

File: index.js [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+  // Combine member lists and remove duplicate members.
+  // Filter out members that are already in target channel.
+  // This prevents cants_invite_self and already_in_channel errors
+  const membersToInvite = [...new Set([...sourceMembers, ...targetMembers])]
+    .filter(member => !targetMembers.includes(member))
+
+  // Slack API limits the number of invites.
+  // Make a new request 20 entries at a time while members remain in the array.
+  while (membersToInvite.length > 0) {
+    const batch = membersToInvite.splice(0, 20);
```
Comment (phoenixcoder) [inline index.js:101]: The max on this is a magic number. You need to move that 20 to constant at the top and call it batch size.

File: index.js [hunk 18]
```diff
@@ -0,0 +1,109 @@
...
+
+    // result currently unused--possible future use.
+    const result = await axios.post(`https://slack.com/api/conversations.invite?token=${accessToken}&channel=${targetId}&users=${batch.join('%2C')}`)
```
Comment (phoenixcoder) [inline index.js:104]: Move URL string to a constant.

Comment (phoenixcoder) [inline index.js:104]: Not sure the point of this await...stop hitting everything with the same hammer, lol :-P

**Review by phoenixcoder — CHANGES_REQUESTED** [review]
Lots of things to address in this one.

File: index.js (hunk 8 above)
Comment (bhurstGH) [reply index.js:25]: Good call, don't disagree.

File: index.js (hunk 9 above)
Comment (bhurstGH) [reply index.js:28]: Can do -- believe that's what it was originally, until I switched over to the Conversation ID convention, and considered it could be used to migrate private convos down the line as well.

File: index.js (hunk 10 above)
Comment (bhurstGH) [reply index.js:33]: Same reasoning -- Conversation API/private convos, but can do.

File: index.js (hunk 9 above)
Comment (phoenixcoder) [reply index.js:28]: @bhurstGH in the response, this is actually called a channel: https://api.slack.com/methods/conversations.list

So, I'm not sure what convention you're referring to.

File: index.js (hunk 12 above)
Comment (phoenixcoder) [reply index.js:39]: Also, just fyi, just because you CAN do recursion, doesn't mean you should. You're filling up your callstack unnecessarily here with every function call. There's a way to do this in a do-while loop.

Comment (bhurstGH) [reply index.js:39]: Not sure I follow here. 

The check is `If the array isn't empty` (which indicates we've gathered some members -- it's not enough if its just the original empty array existing), and `There is no next cursor provided` (which indicates there are no more pages to check), then we've paged through all the results and gotten all the members, so stop recursion and return them.

Also not sure why base case wouldn't trigger under that circumstance -- your the math guy, but isn't that the same as saying totalNumberOfMembers === pageSize? 200 members, default page size is 200, remainder 0. In that case, membersCollection has length, and no nextCursor is provided -- why wouldn't it trigger?

File: index.js (hunk 15 above)
Comment (bhurstGH) [reply index.js:59]: This is where I should have known about `encodeURIComponent()` -- but essentially does the same thing.

As far as I could tell, next_cursor always ended with an equal sign, `=`, which needed to be encoded in the URL. It is in fact broke if I don't do this -- but maybe I missed something? I thought it was kind of odd as well.

File: index.js (hunk 12 above)
Comment (phoenixcoder) [reply index.js:39]: @bhurstGH for this question, I'm going to suggest writing unit test cases for this to help you answer your questions. Basically, nextCursor being empty is sufficient to determine whether we should stop processing.

File: index.js (hunk 16 above)
Comment (bhurstGH) [reply index.js:90]: Ah, I see what you're saying. 

There is a better way regarding concurrency, yeah -- would have to refresh  my memory on best practice there -- but the actual purpose of awaiting at all here is for the code below, so `getConversationMembers()` doesn't run until the IDs have returned, and so we don't filter into `membersToInvite` until the member lists have returned.

File: index.js (hunk 12 above)
Comment (phoenixcoder) [reply index.js:39]: re. member list size...and so is 0 mod pageSize === 0, if Slack ever sends that back. They probably don't, but why risk it when the other check is extraneous?

File: index.js (hunk 9 above)
Comment (bhurstGH) [reply index.js:28]: The *Conversation* API is the convention i was referring to. From your link:

> This Conversations API method returns a list of all channel-like conversations in a workspace. The "channels" returned depend on what the calling token has access to and the directives placed in the types parameter.

And the available types of conversations this could be used include public and private channels, as well as IM and "MPIM" -- which I guess is multi-person IM? Not sure.

No issue with changing it to channels in this context, but that was the logic: the use of the Conversation API and that the possible user lists could come from private messages in addition to channels.

File: index.js (hunk 16 above)
Comment (phoenixcoder) [reply index.js:90]: @bhurstGH Yeah, your membersToInvite is your synchronization point, but the above four should be able to run concurrently...they don't depend on one another.

File: index.js (hunk 15 above)
Comment (phoenixcoder) [reply index.js:59]: @bhurstGH good you found that encodeURIComponent() function. You rarely want to be hand-editing these things yourself.

File: index.js (hunk 18 above)
Comment (bhurstGH) [reply index.js:104]: Since `result` is currently unused, I think my thought was similar to above -- if code below this was to execute using `result`, you don't want to do it until the result has made it back. VS chaining the result into `axios.post().then(result).catch(error)`.

File: index.js (hunk 12 above)
Comment (phoenixcoder) [reply index.js:39]: Talked with @bhurstGH offline. We agreed this could use a refactor, but we'll save it for when this thing moves into a Slack App. It won't help anything to refactor and productionalize this right now without a good test method.

File: index.js (hunk 17 above)
Comment (bhurstGH) [reply index.js:101]: It's been a while... But this MAY have been based on reading something about how `conversation.invite` invites members 20 at a time... Doesn't really effect moving it to a constant, just trying to remember my own reasoning.

File: index.js (hunk 16 above)
Comment (phoenixcoder) [reply index.js:90]: @bhurstGH and I agree to save this for the migration to a Slack App.

File: index.js (hunk 17 above)
Comment (phoenixcoder) [reply index.js:101]: We'll make improvements for the Slack App migration.

File: index.js (hunk 18 above)
Comment (phoenixcoder) [reply index.js:104]: We'll make improvements for the Slack App migration.

Comment (phoenixcoder) [reply index.js:104]: We'll make improvements for the Slack App migration.

File: index.js (hunk 15 above)
Comment (phoenixcoder) [reply index.js:59]: We'll make improvements for the Slack App migration.

File: index.js (hunk 11 above)
Comment (phoenixcoder) [reply index.js:36]: We'll make improvements for the Slack App migration.

File: index.js (hunk 10 above)
Comment (phoenixcoder) [reply index.js:33]: We'll make improvements for the Slack App migration.

File: index.js (hunk 9 above)
Comment (phoenixcoder) [reply index.js:28]: We'll make improvements for the Slack App migration.

File: index.js (hunk 8 above)
Comment (phoenixcoder) [reply index.js:25]: We'll make improvements for the Slack App migration.

File: index.js (hunk 13 above)
Comment (phoenixcoder) [reply index.js:45]: We'll make improvements for the Slack App migration.

File: index.js (hunk 14 above)
Comment (phoenixcoder) [reply index.js:63]: We'll make improvements for the Slack App migration.

File: index.js (hunk 6 above)
Comment (phoenixcoder) [reply index.js:55]: We'll make improvements for the Slack App migration.
